python/ray/data/dataset.py

```diff
@@ -349,14 +350,33 @@ def random_shuffle(self,
         Returns:
             The shuffled dataset.
         """
+        # Check for spread resource labels in environment variable, and use
+        # the given labels for round-robin node-based scheduling.
+        shuffle_spread_custom_resource_labels = os.getenv(
+            "RAY_DATASETS_SHUFFLE_SPREAD_CUSTOM_RESOURCE_LABELS", None)
+        if shuffle_spread_custom_resource_labels is not None:
+            shuffle_spread_custom_resource_labels = (
+                shuffle_spread_custom_resource_labels.split(","))
+            round_robin_resource_provider = itertools.cycle(
+                map(lambda resource: {resource: 0.001},
+                    shuffle_spread_custom_resource_labels))
+        else:
+            round_robin_resource_provider = None
 
+        num_blocks = self.num_blocks()
         new_blocks = simple_shuffle(
-            self._blocks,
-            num_blocks or self.num_blocks(),
+            self.move_blocks(),
+            num_blocks or num_blocks,
             random_shuffle=True,
-            random_seed=seed)
+            random_seed=seed,
+            _round_robin_resource_provider=round_robin_resource_provider)
         return Dataset(new_blocks)
 
+    def move_blocks(self):
+        blocks = self._blocks
```

Comment:
Make this private. Also, can we raise a good error message if self.blocks is None and we try to use the dataset?